src/som/interpreter/nodes/literals/ArrayLiteralNode.java

```diff
@@ -0,0 +1,81 @@
+package som.interpreter.nodes.literals;
+
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.nodes.ExplodeLoop;
+import com.oracle.truffle.api.source.SourceSection;
+
+import som.interpreter.nodes.ExpressionNode;
+import som.vm.constants.Classes;
+import som.vmobjects.SArray.SMutableArray;
+
+public class ArrayLiteralNode extends LiteralNode {
+  @Children final ExpressionNode[] expressions;
+
+  public ArrayLiteralNode(final ExpressionNode[] expressions, final SourceSection source) {
+    super(source);
+    assert source != null;
+    this.expressions = expressions;
+  }
+
+  @ExplodeLoop
+  private Object[] evaluateAsArray(final VirtualFrame frame) {
+    Object[] evaluated = new Object [expressions.length];
+    for (int i = 0; i < expressions.length; i++) {
+      evaluated[i] = expressions[i].executeGeneric(frame);
+    }
+    return evaluated;
+  }
+
+  @ExplodeLoop
+  private boolean headSameTypeAsTail(final Object[] evaluated) {
+    for (int i = 1; i < evaluated.length; i++) {
+      if (!(evaluated[i].getClass().equals(evaluated[0].getClass()))) {
```

Comment:
Comparing the classes here does unfortunately not really work, since we have a wide range of classes floating around, not just primitives and objects.